samples/src/main/java/okio/samples/Throttler.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio.samples;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.concurrent.TimeUnit;
+import okio.Buffer;
+import okio.ForwardingSink;
+import okio.ForwardingSource;
+import okio.Sink;
+import okio.Source;
+
+/**
+ * Enables throttling of Sources and Sinks. Attach to this throttler via {@link #source(Source)} and
+ * {@link #sink(Sink)} and set the desired throughput via {@link #bytesPerTimePeriod(long, long,
+ * TimeUnit)}. Multiple Sources and Sinks can be attached to a single Throttler and they will be
+ * throttled as a group, where their combined throughput will not exceed the desired throughput. The
+ * same Source or Sink can be attached to multiple Throttlers and its throughput will not exceed the
+ * desired throughput of any of the Throttlers.
+ *
+ * <p>To clear any throttling, use {@link #clear()} or specify a <code>byteCount</code> of <code>
+ * 0</code> when setting the throughput.
+ */
+public final class Throttler {
+  private static final int NANOS_PER_MILLI = 1000000;
+
+  private double limit = 0.0d;
+  private long last = -1;
+
+  /** Sets the desired max throughput. <code>byteCount</code> of zero to clear. */
+  public void bytesPerTimePeriod(long byteCount, long duration, TimeUnit unit) {
+    if (unit == null || byteCount < 0 || duration <= 0) {
+      throw new IllegalArgumentException(
+          "byteCount=" + byteCount + " duration=" + duration + " unit=" + unit);
+    }
+
+    synchronized (this) {
+      this.limit = (double) byteCount / unit.toNanos(duration);
+    }
+  }
+
+  /** Clears the max throughput. */
+  public void clear() {
+    synchronized (this) {
+      this.limit = 0.0d;
```

Comment:
notifyAll ?

Reply:
There shouldn't be a need. This is holding the synchronized lock so no Sources or Sinks should be reading or writing.

Reply:
Ahh. . .  you wanna change `delay()` to call `wait()` instead of `Thread.sleep()` because the former can be woken up early with a call to `notifyAll()`. And you wanna wake it up early if more bandwidth becomes available while it’s sleeping.

Reply:
Using `wait()` actually breaks the entire implementation because it would release the synchronization lock. I want the Threads to be stuck on the synchronization, I'm using it like a queue. All requests will get queued by the synchronization block and each request will be delayed as much as necessary depending on how long ago the last request was. 

In terms of the Leaky Bucket algorithm, think of the synchronization lock as the bucket and the leak is the `sleep()` and the `last` variable is the inverse of how full the bucket is.

I'm hoping the graphs that I created in the `delay()` method shed some light on how the implementation works in practice. Maybe I can get more detailed or add some descriptions to each example.